roachtest: sqlsmith/setup=empty/setting=no-mutations failed

[(roachtest).sqlsmith/setup=empty/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2670113&tab=buildLog) on [master@e9e372122a2e3db7090b5705da07128f828e2441](https://github.com/cockroachdb/cockroach/commits/e9e372122a2e3db7090b5705da07128f828e2441):

```
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(
									'BOX(-1.5744801441378615 0.05490464231675818,-0.4702548486709238 2.02313088893291)':::BOX2D,
									(-2052378172):::INT8,
									NULL
								),
								(
									'BOX(-1.4422313710676344 -0.05568743776837182,0.20284876008875963 0.1097720387790686)':::BOX2D,
									(-841540906):::INT8,
									NULL
								),
								(
									'BOX(-1.2627687118839002 -0.1297262421243775,2.1042446892636475 -0.06826572975969625)':::BOX2D,
									350429516:::INT8,
									NULL
								),
								('BOX(-10 -10,10 10)':::BOX2D, NULL, 'a73711ac-599e-4dc2-bad1-a0eba2d7b2f2':::UUID)
						)
							AS tab_18254 (col_31656, col_31657, col_31658)
				),
			with_5764 (col_31659)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								('2d96a727-6c71-4ace-894b-a7a7fc9dce8e':::UUID),
								('ec500b14-aff4-4a9f-b6b9-31777bd77b19':::UUID),
								('ab0a5fe3-8ed4-41a8-ab6f-9b9882c964ba':::UUID)
						)
							AS tab_18255 (col_31659)
				)
		SELECT
			cte_ref_1727.col_31659 AS col_31660
		FROM
			with_5764 AS cte_ref_1727
		GROUP BY
			cte_ref_1727.col_31659
		HAVING
			_st_containsproperly(st_mlinefromtext('*5':::STRING::STRING, crdb_internal.complete_stream_ingestion_job((-7819254773292031758):::INT8::INT8)::INT8::INT8)::GEOMETRY::GEOMETRY, NULL::GEOMETRY)::BOOL
		ORDER BY
			cte_ref_1727.col_31659
		LIMIT
			17:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2670113&tab=artifacts#/sqlsmith/setup=empty/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=empty/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2676655&tab=buildLog) on [master@5971ecb9dd1a25c81cd6012d6be1ff922802eae5](https://github.com/cockroachdb/cockroach/commits/5971ecb9dd1a25c81cd6012d6be1ff922802eae5):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
	sqlsmith.go:198,sqlsmith.go:228,test_runner.go:767: error: pq: internal error: runtime error: invalid memory address or nil pointer dereference
		stmt:
		SELECT
			e'{"0f#": [], "9u `m6lhQ": 0.8889105461246096, "foobar": [[[], {"Fg>+{pR[nE": {"c": "bXNC$Zg: f", "foo": {}}, "o\\\\\\"I\'`RU}": "baz"}, false], {"I:.k_I<zsA": true}]}':::JSONB
				AS col_1434
		FROM
			(
				VALUES
					('{"baz": {"bar": null, "foobar": null, "z=w/qc": false}}':::JSONB),
					(
						'[{"a": {"bar": {"baz": []}, "qrNWdf": true}, "baz": false, "n]79Y]h!": "baz"}, "baz", [true], 0.197091753371324]':::JSONB
					)
			)
				AS tab_849 (col_1433)
		LIMIT
			49:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2676655&tab=artifacts#/sqlsmith/setup=empty/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/opt/norm/project_funcs.go
+++ b/pkg/sql/opt/norm/project_funcs.go
@@ -365,6 +365,11 @@ func (c *CustomFuncs) CanUnnestJSONFromValues(
 			return false
 		}
 		currJSON := expr.(*memo.ConstExpr).Value.(*tree.DJSON)
+		if currJSON.Type() != json.ObjectJSONType {
+			// This value is not an object. It is important to check, because a JSON
+			// array can pass the checks below (see #60522).
+			return false
+		}
 		iter, err := firstJSON.ObjectIter()
 		if err != nil {
 			return false

--- a/pkg/sql/opt/norm/testdata/rules/project
+++ b/pkg/sql/opt/norm/testdata/rules/project
@@ -1478,3 +1478,20 @@ project
  │    └── ('{"y": "three"}',)
  └── projections
       └── column1:1::JSONB->'x' [as=x:2, outer=(1), immutable]
+
+# Verify that the rule doesn't fire because of an array that has the right key
+# as an element (#60522).
+norm expect-not=FoldJSONAccessIntoValues
+SELECT j->'foo' FROM (VALUES ('{"foo": "bar"}'::JSONB), ('["foo", "baz"]'::JSONB)) AS v(j)
+----
+project
+ ├── columns: "?column?":2
+ ├── cardinality: [2 - 2]
+ ├── immutable
+ ├── values
+ │    ├── columns: column1:1!null
+ │    ├── cardinality: [2 - 2]
+ │    ├── ('{"foo": "bar"}',)
+ │    └── ('["foo", "baz"]',)
+ └── projections
+      └── column1:1->'foo' [as="?column?":2, outer=(1), immutable]

--- a/pkg/util/json/json.go
+++ b/pkg/util/json/json.go
@@ -140,7 +140,11 @@ type JSON interface {
 	// type, and a boolean inidicating if this JSON document is a numeric type.
 	AsDecimal() (*apd.Decimal, bool)
 
-	// Exists implements the `?` operator.
+	// Exists implements the `?` operator: does the string exist as a top-level
+	// key within the JSON value?
+	//
+	// If the object is a JSON array, returns true when the key is a top-level
+	// element of the array.
 	Exists(string) (bool, error)
 
 	// StripNulls returns the JSON document with all object fields that have null values omitted

